### PR TITLE
feat(DatePicker): Convert to functional components with typescript and hooks

### PR DESCRIPTION
**Pull request by joakbjerk**
- [x] `DatePicker`
- [x] `DatePickerAddon`
- [x] `DatePickerCalc`
- [x] `DatePickerCalendar`
- [x] `DatePickerContext`
- [x] `DatePickerFooter`
- [x] `DatePickerInput`
- [x] `DatePickerProvider`
- [x] `DatePickerRange`
- [x] Fix tests
- [x] Fix types
- [x] Fix `onClickOutside` functionality
- [x] Make tests for `onClickOutside`
- [x] Fix `DatePickerProvider` `getDerivedStateFromProps` logic
- [x] Fix `DatePickerInput` mask and value logic
- [x] Fix `DatePickerInput` paste functionality
- [x] General Bugfixes
- [x] Move provider date logic to hook
- [x] Move provider view logic to hook
- [x] Move date correction logic closer to trigger
- [x] Move `lastEventCallCache` to hook
- [x] Fix `callOnChangeHandler` callback functionality
- [x] Fix Input onChange handling
- [x] Wrap necessary functions or values in `useCallback` or `useMemo` hooks
- [x] Remove `locale` prop, and depricate (Solved by only importing the two relevant functions used by the `Locale` object)
- [x] Replace all `any` types with more accurate ones
- [x] Fix Range Keyboard functionality
- [x] Clean up mess 🙏
- [x] fix type errors
- [x] Fix shortcuts
- [x] Fix cancel button
- [x] Stop cancel button from resetting date
- [x] Fix latest discrepancies
- [x] Figure out the `DatePickerEvent` situation
- [x] Rename `ReturnObject`-types to give better context
- [x] Fix Input date removal when max or min is active
- [x] Add tests range shortcuts
- [x] Add tests for `link` selection
- [x] Add tests for range pickers cancel button behaviour
- [x] Add tests to make sure shortcut callback receives all dates as parameters
- [x] Add tests for `endDate` backspace removal when in `range`
- [x] Update month handling
- [x] Add more tests for range view functionality
- [x] Fix clicking same date for end and start date while in range mode, and add test
- [x] Contemplate life 🤔
- [x] Fix range selection of start and end date in same month behaviour
- [x] Fix date selection with arrow keys and add test

### CODE REVIEW

**Comment by joakbjerk** [comment]
> Great job, so many changes and commits 🤯 
> Do you want me to do some testing of the current state of this PR, @joakbjerk ?

Yes please, if you have the time ❤️

**Comment by langz** [comment]
Datepicker does not update value when pressing "today"-button.

Video from preview of this PR:

https://github.com/dnbexperience/eufemia/assets/1359205/2148ece3-f334-4002-bf26-fd37f61f889b



https://github.com/dnbexperience/eufemia/assets/1359205/1a839f6f-15fe-495c-82a4-6c4b7c2beb66


Compare the behavior of https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#range-datepicker and https://eufemia.dnb.no/uilib/components/date-picker/#range-datepicker 


Reproduce:
- Go to https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#range-datepicker
- Click datepicker icon
- Click today
- Input(values) does not change/update. It does so in [main branch](https://eufemia.dnb.no/uilib/components/date-picker/#range-datepicker) today.

Expected behavior(video from main):

https://github.com/dnbexperience/eufemia/assets/1359205/1a839f6f-15fe-495c-82a4-6c4b7c2beb66

**Comment by langz** [comment]
Avbryt/Cancel-button doesn't seem like it's working at all.

Compare the [preview of this PR](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#range-datepicker) vs [main](https://eufemia.dnb.no/uilib/components/date-picker/#range-datepicker):

When pressing "Avbryt"-button nothing happens, or the modal/datepicker doesn't close as expected:

https://github.com/dnbexperience/eufemia/assets/1359205/0167e036-cd6e-4155-a062-12b94c7485e0

**Comment by langz** [comment]
Avbryt/Cancel-button when pressed in the following example changes the selected dates/values in the datepicker.
It's probably related to this [comment](https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2194337635).

Compare [main](https://eufemia.dnb.no/uilib/components/date-picker/#opened-datepicker) vs [preview of this PR](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#opened-datepicker), click the "avbryt"-button:


https://github.com/dnbexperience/eufemia/assets/1359205/2b16f5ca-0789-4d6d-9537-a4602697ed33

Expected behavior(from main):

https://github.com/dnbexperience/eufemia/assets/1359205/dd6ff7fb-7ffe-464d-9270-caaf1674ccfc

**Comment by langz** [comment]
Just an observation, not sure if this is an actual problem.
I've tabbed into the datepicker, then entered the following(0 1 0 1 2 0 2 2) and this is the output I get from main:

and this is the output I get from the deploy preview:

```
on_change 2024-06-02
on_change 2024-06-01
onBlur 2024-06-01
on_change 2024-01-01
onBlur 2024-01-01
on_change 2202-01-01
on_change 2020-01-01
on_change 2022-01-01
```


Output from main:

```
on_change 2024-06-02
on_change 2024-06-01
onBlur 2024-06-01
on_change null
on_change 2024-01-01
onBlur 2024-01-01
on_change 2202-01-01
on_change 2020-01-01
on_change 2022-01-01
```



The output is not identical, even though my input is the same, but it doesn't necessarily mean that "it's not working".
I actually think the output now(in this PR) is better than on main.

**Comment by langz** [comment]
Date displayed is

Screenshot 2024-10-01 at 13 51 15

**Comment by langz** [comment]
The [Linked DatePickers](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#linked-datepickers) example does not work in the same manner in [main branch](https://eufemia.dnb.no/uilib/components/date-picker/#linked-datepickers) as in [this deploy preview](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#linked-datepickers):


https://github.com/user-attachments/assets/f64c8b7f-38a5-435c-8868-58aae3c64654

In the deploy preview, when selecting a date in november the datepicker changes/updates the datepicker to display november and december while in main branch it doesn't change/update and still displays oktober and november.






In the screenshot above, you can see main branch on the left side and the deploy preview of this PR on the right side.

**Review by langz — COMMENTED** [review]
Looks much better now, thanks for improving 🚀 
I observed the following "concerns":
- https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2385642232
- https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2385582967

**Comment by joakbjerk** [comment]
> Looks much better now, thanks for improving 🚀 I observed the following "concerns":
> 
>     * [feat(DatePicker): Convert to functional components with typescript and hooks #2799 (comment)](https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2385642232)
> 
>     * [feat(DatePicker): Convert to functional components with typescript and hooks #2799 (comment)](https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2385582967)

Sweet 🔥

**Comment by joakbjerk** [comment]
> The "relative + 3 days"-button does not behave correctly(in the same manner as on main branch):
> Untitled.mov
> 
> In the screenshot above, you can see main branch on the left side and the deploy preview of this PR on the right side.

This one can be retested @langz

**Comment by langz** [comment]
> > The "relative + 3 days"-button does not behave correctly(in the same manner as on main branch):
> > Untitled.mov
> > In the screenshot above, you can see main branch on the left side and the deploy preview of this PR on the right side.
> 
> This one can be retested @langz

This looks to work great 👍

**Comment by joakbjerk** [comment]
> The [Linked DatePickers](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#linked-datepickers) example does not work in the same manner in [main branch](https://eufemia.dnb.no/uilib/components/date-picker/#linked-datepickers) as in [this deploy preview](https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/#linked-datepickers):
> Untitled2.mov
> 
> In the screenshot above, you can see main branch on the left side and the deploy preview of this PR on the right side.
> 
> In the deploy preview, when selecting a date in november the datepicker changes/updates the datepicker to display november and december while in main branch it doesn't change/update and still displays oktober and november.
> 
> This happens in a lot of other examples as well, where using `range` property.
> 
> Additionally, take a look at this example when selecting 01.10.2024 - 02.10.2024. It also behaves differently on main and in deploy preview:
> unt3.mov
> 
> In the screenshot above, you can see main branch on the left side and the deploy preview of this PR on the right side.

This should be fixed now @langz 🕺

**Comment by joakbjerk** [comment]
This fun bug needs to be fixed too 😾

https://github.com/user-attachments/assets/27fc3c61-7592-4787-aabd-85f13ebb7a39

**Comment by langz** [comment]
Something magical:

https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#default-datepicker

- Open datepicker
- Click left arrow(change to April)
- Then hover on 7
- It magically updates/changes to display the date(selected date) 5th of may

**Comment by langz** [comment]
Linked datePickers.
Try to compare the two https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers vs. https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers

In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
- Press back arrow to september, then select september 4th
- Datepicker magically changes to display october.

This behavior doesn't happen in https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers

**Comment by joakbjerk** [comment]
Both the above issues should be fixed with the latest commit 🤞 @langz

**Comment by langz** [comment]
> Linked datePickers. Try to compare the two https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers vs. https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers
> 
> In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
> 
> * Press back arrow to september, then select september 4th
> * Datepicker magically changes to display october.
> 
>  Screen.Recording.2024-10-23.at.13.55.32.mov 
> This behavior doesn't happen in https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers

This does now happen to the "right date picker".
In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:

- Select 01.10.2024
- Select 01.10.2024
- Datepicker magically changes the right datepicker to display october.

I would expect the right datepicker to still show november, as it does in main branch.

**Comment by langz** [comment]
❤️ Congratulations to the 1 year anniversary of this PR ❤️ 

image


On a serious note, thanks for the great work @joakbjerk 🙇

**Comment by joakbjerk** [comment]
> > Linked datePickers. Try to compare the two https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers vs. https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers
> > In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
> > 
> > * Press back arrow to september, then select september 4th
> > * Datepicker magically changes to display october.
> > 
> > Screen.Recording.2024-10-23.at.13.55.32.mov
> > This behavior doesn't happen in https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers
> 
> This does now happen to the "right date picker". In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
> 
>     * Select 01.10.2024
> 
>     * Select 01.10.2024
> 
>     * Datepicker magically changes the right datepicker to display october.
> 
> 
> Screen.Recording.2024-10-25.at.10.51.53.mov
> 
> I would expect the right datepicker to still show november, as it does in main branch.
> 
> Here's how the same actions look in main branch:
> Screen.Recording.2024-10-25.at.10.52.31.mov
> 
> You can also experience this issue when selecting a date in may twice in the following example.
> 
> https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#opened-datepicker
> Screen.Recording.2024-10-25.at.11.00.24.mov

Fix and added a test for this now @langz

**Comment by langz** [comment]
> > > Linked datePickers. Try to compare the two https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers vs. https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers
> 
> > > In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
> 
> > > 
> 
> > > * Press back arrow to september, then select september 4th
> 
> > > * Datepicker magically changes to display october.
> 
> > > 
> 
> > > Screen.Recording.2024-10-23.at.13.55.32.mov
> 
> > > This behavior doesn't happen in https://eufemia.dnb.no/uilib/components/date-picker/demos/#linked-datepickers
> 
> > 
> 
> > This does now happen to the "right date picker". In https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#linked-datepickers:
> 
> > 
> 
> >     * Select 01.10.2024
> 
> > 
> 
> >     * Select 01.10.2024
> 
> > 
> 
> >     * Datepicker magically changes the right datepicker to display october.
> 
> > 
> 
> > 
> 
> > Screen.Recording.2024-10-25.at.10.51.53.mov
> 
> > 
> 
> > I would expect the right datepicker to still show november, as it does in main branch.
> 
> > 
> 
> > Here's how the same actions look in main branch:
> 
> > Screen.Recording.2024-10-25.at.10.52.31.mov
> 
> > 
> 
> > You can also experience this issue when selecting a date in may twice in the following example.
> 
> > 
> 
> > https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/components/date-picker/demos/#opened-datepicker
> 
> > Screen.Recording.2024-10-25.at.11.00.24.mov
> 
> 
> 
> Fix and added a test for this now @langz

I'll do one last round of testing today❤️

**Comment by langz** [comment]
# Field.Date

When testing `Field.Date`(comparing https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/extensions/forms/feature-fields/Date/demos/#empty vs. https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/Date/demos/#empty), using arrow keys on my keyboard to select the date in the datepicker doesn't work as expected(I have to press the arrow twice for the selected date to move/change):


https://github.com/user-attachments/assets/66fb555b-3d45-4881-8e42-1840f346c2cb



I've only experienced this in `Field.Date`, and not in the `DatePicker`:

**Comment by langz** [comment]
After testing this sometimes now, all that's left is this [issue](https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2441472714), which I also think we can solve in a separate PR. Therefore, I'm approving this PR now ✅ 

@tujoworker, do you have time to take a look at the code in this PR? 🙏

File: packages/dnb-design-system-portal/src/docs/uilib/components/date-picker/properties.mdx [hunk 1]
```diff
@@ -35,7 +35,6 @@ import {
...
 | `first_day`                             | _(optional)_ to define the first day of the week. Defaults to `monday`.                                                                                                                                                                                                                                                                                                              |
-| `locale`                                | _(optional)_ to define the locale used in the calendar. Needs to be an `date-fns` **v2** locale object, like `import enLocale from 'date-fns/locale/en-GB'`. Defaults to `nb-NO`.                                                                                                                                                                                                    |
```
Comment (langz) [inline packages/dnb-design-system-portal/src/docs/uilib/components/date-picker/properties.mdx:38]: Is this a breaking change, that we no longer provide the `locale` property?
If someone is using this property right now, is it safe to just remove it? 
Should we mention the removal(and/or what users should do to keep the same functionality, if it's stille possible?)  in some kind of release notes or just docs in `DatePicker`?

**Review by langz — APPROVED** [review]
After testing this sometimes now, all that's left is this [issue](https://github.com/dnbexperience/eufemia/pull/2799#issuecomment-2441472714), which I also think we can solve in a separate PR. Therefore, I'm approving this PR now ✅ 

@tujoworker, do you have time to take a look at the code in this PR? 🙏 

EDIT: Issue is fixed now🤌

File: packages/dnb-design-system-portal/src/docs/uilib/components/date-picker/properties.mdx (hunk 1 above)
Comment (joakbjerk) [reply packages/dnb-design-system-portal/src/docs/uilib/components/date-picker/properties.mdx:38]: I think this was replaced by a locale/translation rewrite we did earlier this year. Cant remember the exact details though

**Comment by joakbjerk** [comment]
> # Field.Date
> 
> When testing `Field.Date`(comparing https://eufemia-git-chore-datepicker-ts-eufemia.vercel.app/uilib/extensions/forms/feature-fields/Date/demos/#empty vs. https://eufemia.dnb.no/uilib/extensions/forms/feature-fields/Date/demos/#empty), using arrow keys on my keyboard to select the date in the datepicker doesn't work as expected(I have to press the arrow twice for the selected date to move/change):
> Screen.Recording.2024-10-28.at.13.36.15.mov
> 
> I've only experienced this in `Field.Date` which use the `DatePicker`, and not in the actual `DatePicker`.

This issue is fixed now 🙏 @langz

**Review by tujoworker — APPROVED** [review]
Fantastic 🕺

**Pull request merged**